Add a fix language function

JeffreyWangDev opened the pull request:
Added a function to rename the language name to consolidate the same language, e.g., js and javascript to JavaScript. 

Bug report: https://github.com/hackclub/hackatime/issues/402

JeffreyWangDev committed:
Add a fix language function

Added a function to rename the language name to consolidate the same language EX js and javascript to Javascript.
Bug report: https://github.com/hackclub/hackatime/issues/402

JeffreyWangDev committed:
Remove white space

I literally removed 2 lines of nothing bc the linter was mad

Hopefully the linter is happy now

JeffreyWangDev committed:
Remove 2 tabs

I removed 2 tabs from line 135

The linter was still mad, I hope to all powerful that its happy now

JeffreyWangDev commented:
Note: This does not fix any old heartbeats. I don't know how/where to write a script to do so.

JeffreyWangDev commented:
Used this to test out this function


```
$time = [Math]::Floor([decimal](Get-Date(Get-Date).ToUniversalTime()-uformat '%s'))
$heartbeat = @{
    type = 'coding'
    time = $time
    entity = 'test.js'
    language = 'js'
    project = 'test-project'
}

$response = Invoke-RestMethod -Uri "http://localhost:3000/api/hackatime/v1/users/current/heartbeats" `
    -Method Post `
    -Headers @{Authorization="Bearer ea23e393-574c-454f-8927-a4584ede0bda"} `
    -ContentType 'application/json' `
    -Body "[$($heartbeat | ConvertTo-Json)]"
    
$heartbeat = @{
    type = 'coding'
    time = $time
    entity = 'test.js'
    language = 'JavaSCRIPT'
    project = 'test-project'
}
$response = Invoke-RestMethod -Uri "http://localhost:3000/api/hackatime/v1/users/current/heartbeats" `
    -Method Post `
    -Headers @{Authorization="Bearer ea23e393-574c-454f-8927-a4584ede0bda"} `
    -ContentType 'application/json' `
    -Body "[$($heartbeat | ConvertTo-Json)]"
```

It worked on my side with JS and Java

<img width="1209" height="644" alt="image" src="https://github.com/user-attachments/assets/41a37f07-26f1-4cef-b492-344fc590bd10" />

maxwofford commented:
I like the initial approach, but this is going to take a little bit more thinking. We need to set this up in a way where we don't clobber any data. The raw fields being sent to us shouldn't be rewritten without backing them up!

JeffreyWangDev commented:
Il change it so it does the renaming on the UI side of it, would that be better?

maxwofford commented:
@JeffreyWangDev if you can get away with that, sure! I imagine it's going to be hard to get everything though because there are some more complex things like the graphs that generate on the frontend based on SQL queries. Try it out and see how possible it is.

The other option is something like a "raw_language" and "language" field- the "raw_language" is what comes in originally, the "language" is the one we remap

JeffreyWangDev committed:
Add function to rename languages

Added a function to strings to rename the language to a universal name in the main home page

JeffreyWangDev commented:
Idk if it works in other places cause I didnt put it in other places. I don't really have access to other places like the admin pages.

JeffreyWangDev committed:
Remove whitespace

I removed 3 spaces bc linter was mad

JeffreyWangDev committed:
Add name checking in filter

Fixed the filters working with the new name thing

JeffreyWangDev commented:
I got the graphs and the dropdown to reduce languages by renaming it in the heartbeats

JeffreyWangDev commented:
https://github.com/hackclub/hackatime/issues/358

3kh0 commented:
will review shortly

ok-coder1 commented:
> will review shortly

"shortly". Please check again.

JeffreyWangDev commented:
kinda forgot this existed, please let me know what to change

JeffreyWangDev commented:
@3kh0 bump?

3kh0 reviewed (changes requested):
Conflicts present, resolve and I will review

JeffreyWangDev committed:
Merge branch 'main' into Name-capitalization

JeffreyWangDev commented:
I fixed it with the commit

3kh0 committed:
Merge branch 'main' into Name-capitalization

JeffreyWangDev commented on app/controllers/static_pages_controller.rb:
```diff
@@ -45,8 +45,10 @@ def index
+          .map { |r| [ r.language&.categorize_language, r.language_count ] } # fix the bug where langs can have both upper and lower case like JAVA and java found here (https://github.com/hackclub/hackatime/issues/402)
           .reject { |lang, _| lang.nil? || lang.empty? }
+          .group_by { |lang, _| lang }
+          .transform_values { |pairs| pairs.sum { |_, count| count } }
```
where???

3kh0 committed:
Merge branch 'main' into Name-capitalization
